activesupport/lib/active_support/cache/key.rb

```diff
@@ -0,0 +1,161 @@
+    # that can be implicitly cast to a string:
+    #
+    # Examples:
+    #
+    #   Key.new("foo").cache_key
+    #   # => "foo"
+    #
+    #   Key.new(["foo", "bar"]).cache_key
+    #   # => "foo/bar"
+    #
+    #   Key.new({ bar: 1, foo: 2 }).cache_key
+    #   # => "bar=1/foo=2"
+    #
+    #   klass = Class.new { def cache_key; "foo"; end }
+    #   Key.new(klass.new).cache_key
+    #   # => "foo"
+    #
+    # In addition to generating a `cache_key`,
+    # a `cache_version` is also generated if the object
+    # passed in responds to `cache_version` or if it contains
+    # an element that responds to `cache_version`.
+    #
+    # The purpose of keeping this value seperate from
+    # `cache_key` is to support the use of "cache versioning"
+    # where the same key is recycled but the version of the object
+    # being stored is kept directly in the cache.
+    #
+    # Example:
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new(klass.new).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new([klass.new]).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new({foo: klass.new}).cache_version
+    #   # => "foo"
+    class Key # :nodoc:
+      attr_reader :cache_key, :cache_version
+
+      def self.call(key)
+        if key.is_a?(Key)
+          key
+        else
+          self.new(key)
+        end
+      end
+
+      def initialize(key = nil)
+        @cache_key = +""
+        @cache_version = +""
+        @empty_key = true
+        self << key
+      end
+
+      def update(key)
+        if key.respond_to?(:cache_key)
+          update_version(key)
+          update_key(key.cache_key.to_s)
+        else
+          split_key(key)
+        end
+      end
+      alias :<< :update
+
+      private
+        def update_key(key)
+          @cache_key << "/" unless @empty_key
+          @cache_key << key.to_param.to_s
+          @empty_key = false
+        end
+
+        def update_version(key)
+          if key.respond_to?(:cache_version)
+            @cache_version << "/" unless @cache_version.empty?
+            @cache_version << key.cache_version.to_param.to_s
+          else
+            split_version(key)
+          end
+        end
+
+        def split_key(key)
+          case key
+          when Array
+            key.each { |k| update(k) }
+          when Hash
+            key = key.to_a
+            update_version(key)
+            key.sort_by! { |k, _| k.to_s }
+            key.each { |k, v| update_key("#{k}=#{v}") }
+          else
+            update_version(key)
+            update_key(key)
+          end
+        end
+
+        def split_version(key)
+          if key.is_a?(Array)
+            key.each { |k| update_version(k) }
+          elsif key.respond_to?(:to_a)
+            split_version(key.to_a)
+          end
+        end
+    end
```

Comment:
Implementation is feeling understandable. Still a challenge to grok that it's recursively, incrementally building a serialized key name and version.

Makes me dream of "just" using Digest classes for both! Breaking all compat for sure—but already built for efficient incremental digest generation.

Reply:
Agreed, there are lots of little edge cases here-and-there, where different keys can produce the same key. Overall it seems that's not caused major issues so it seems fine. To make issues worse, the `expand_cache_key` method has a slightly different implementation of how cache key expansion is done.

To add support for the latest dalli release I have to add `cache_key_with_version`. The implementation (not currently pushed) is not fully backwards compatable. For example if there are multiple keys instead of `<cache-key-1>-<cache-version-1>/<cache-key-2>-<cache-version-2>` we end up with `<cache-key-2>/<cache-key-2>-<cache-version-1>/<cache-version-2>`. We can make it backwards compat, but it will require storing and writing a third string.